File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.js
function solution(record) {
    const answer = [];
    const userMap = new Map();
    const commandMap = {"Enter" : "들어왔습니다.", "Leave" : "나갔습니다."};
    
    record.forEach((message) => {
        const [command, id, name] = message.split(" ");
        
        if (command === 'Change' || command === 'Enter') {
            userMap.set(id, name);
        }
        
        if (commandMap[command]) {
            answer.push([id, `님이 ${commandMap[command]}`])
        }
    })
    
    return answer.map((user) => {
        const id = user[0]
        const name = userMap.get(id);
        user[0] = name;
    
        return user.join('');
    })
}

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.*;

public class Solution {
    public String[] solution(String[] record) {
        List<String> answer = new ArrayList<>();
        Map<String, String> userMap = new HashMap<>();

        for (String message : record) {
            String[] components = message.split(" ");
            String command = components[0];
            String id = components[1];

            if (command.equals("Enter") || command.equals("Change")) {
                String name = components[2];
                userMap.put(id, name);
            }
        }

        for (String message : record) {
            String[] components = message.split(" ");
            String command = components[0];
            String id = components[1];

            if (command.equals("Enter")) {
                answer.add(userMap.get(id) + "님이 들어왔습니다.");
            } else if (command.equals("Leave")) {
                answer.add(userMap.get(id) + "님이 나갔습니다.");
            }
        }

        return answer.toArray(new String[0]);
    }
}